Improve dependency management documentation (#556)
AleoHQ/welcome: documentation/guides/02_dependencies.md

---
id: dependencies
title: Dependency Management
sidebar_label: Dependency Management
---
[general tags]: # (guides, dependency, dependency_management, imports, program)

Leo programs can depend on other Aleo programs. There are two types of dependencies:

- **Network dependencies**: Programs already deployed on the Aleo network, fetched as pre-compiled bytecode.
- **Local dependencies**: Leo source code on your filesystem, compiled from source.

## Adding Dependencies

### Network Dependencies

To depend on a program deployed on the Aleo network:
```bash
leo add credits.aleo
```
or
```bash
leo add credits
```

For mainnet programs, specify the network:
```bash
leo add credits --network mainnet

> ```

Author: My instinct is to only show flags when they're necessary, since testnet is the default, `leo add credits` just works. Showing `--network testnet` everywhere might imply it's required when it's not. The mainnet example already demonstrates the flag exists for when you need it. Thoughts?

}
```

### Local Dependencies

> To depend on a Leo package on your filesystem:

Reviewer: ```suggestion
To add a Leo package or Aleo Instructions file from your local filesystem as a dependency:
```

```bash
leo add my_library.aleo --local ./path/to/my_library
```

This records the path in `program.json`:
```json
{
  "dependencies": [
    {
      "name": "my_library.aleo",
      "location": "local",
      "network": null,
      "path": "./path/to/my_library"
    }
  ]
}
```

Local dependencies are compiled from source whenever you build. They never require network access.

### Removing Dependencies
```bash
leo remove credits.aleo
```

## Using Dependencies in Code

> In your `main.leo`, import dependencies before the program declaration:

Reviewer: ```suggestion
In your `main.leo` file, import dependencies before the program declaration:
```

```leo
import credits.aleo;
import my_library.aleo;

program my_program.aleo {
    // ...
}
```

## How Dependency Resolution Works

When you run a Leo command, dependencies are resolved as follows:

1. **Read `program.json`** to find declared dependencies
2. **For each dependency:**

>    - **Local**: Read the Leo source from the specified path and compile it

Reviewer: ```suggestion
   - **Local**: Read the Leo source from the specified path and compile it, or use Aleo Instructions file
```

| `leo upgrade` | Always fetches fresh |
| `leo synthesize` | Always fetches fresh |

Commands that generate proofs (`execute`, `deploy`, `upgrade`, `synthesize`) always fetch fresh bytecode because proofs include commitments to the exact bytecode of dependencies. Using stale cached bytecode would produce invalid proofs if a dependency has been upgraded on-chain.

To force a fresh fetch during `build` or `run`:
```bash
leo build --no-cache
```

## Editions (Program Versions)

An **edition** is the version number of a deployed program on the Aleo network:
- Edition 0: Initial deployment
- Edition 1: First upgrade
- Edition 2: Second upgrade
- ...and so on

By default, Leo fetches the **latest** edition of a network dependency. To pin to a specific edition:
```bash
leo add some_program.aleo --edition 3
```

This records the pinned edition in the manifest:
```json
{
  "name": "some_program.aleo",
  "location": "network",
  "network": "testnet",
  "path": null,
  "edition": 3
}
```

**When to pin editions:**
- When you need reproducible builds
- When a dependency upgrade would break your program
- When you want to avoid unexpected behavior changes

> **Note:** Local dependencies don't have editions - they're always compiled from your current source code.

Reviewer 2: What happens when the local dependency also lives in the network? 

In any case, it must have some edition right, either 0 if this is the first deployment or perhaps higher if it's already present?

Not sure if the edition is actually materially used for anything.
